### PR TITLE
Add regression tests for C++ sample programs

### LINKED ISSUE #774: Fix examples in `samples` folder
Please fill in the following information to report a problem with Cantera.
If you have a question about using Cantera, please post it on our
[Google Users' Group](https://groups.google.com/forum/#!forum/cantera-users).

**System information**

- Cantera version: 2.5.0a3 (master branch)
- OS: Linux
- Python/MATLAB version: N/A

**Expected behavior**

All samples work without errors/warnings.

**Actual behavior**

One sample fails (missing input file)
```
$ ./LiC6_electrode 

***********************************************************************
CanteraError thrown by Application::findInputFile:

Input file LiC6_electrodebulk.xml not found in directories 
'.', 
'/usr/local/share/cantera/data'

To fix this problem, either:
    a) move the missing files into the local directory;
    b) define environment variable CANTERA_DATA to
         point to the directory containing the file.
***********************************************************************
```
and another raises a warning (issue caused by PR #632)
```
$ ./combustor 
DeprecationWarning: FlowDevice::setMassFlowRate(): To be removed after Cantera 2.5. Use device-specific functions (e.g. Valve::setValveCoeff) instead.
DeprecationWarning: Valve::setParameters(): To be removed after Cantera 2.5. Use Valve::setValveCoeff instead.
```

**To Reproduce**

1. Compile cantera from master (e.g. c4c30e3)
2. Run `scons samples`
3. Run examples in `build/samples`

### PR DESCRIPTION
**Pull request by speth**
Similar to the tests added in #874 for the Fortran sample program, this PR adds tests for the C++ samples so they won't break without us knowing.

**Changes proposed in this pull request**

- Add regression tests to run all of the C++ "sample" programs and compare their output against nominal output files
- Remove redundant regression tests which were essentially duplicates of the sample programs
- Add ability to have data files in subdirectories of the Cantera data directory (see Cantera/enhancements#22, #774)
- Make some improvements to test file comparisons for the C++ regression tests
- Fix issues with default composition of `air.yaml` and use of that file in the `combustor.cpp` sample
- Deprecate the argument `ndim` to `newTransportMgr` which is currently ignored / unused
- Fix erroneous deprecation warnings issued by including `onedim.h` (diagnosed thanks to running the `bvp` sample)

**Checklist**

- [X] There is a clear use-case for this code change
- [X] The commit message has a short title & references relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] The pull request is ready for review

**If applicable, fill in the issue number this pull request is fixing**

Fixes #774

### CODE REVIEW

**Comment by ischoegl** [comment]
> Fix erroneous deprecation warnings issued by including onedim.h (diagnosed thanks to running the bvp sample)

Missed that one ... thanks for fixing!

**Comment by speth** [comment]
Rebasing this after merging #886 allows the tests to run reliably on all platforms, so I think this is ready now.

File: samples/cxx/gas_transport/gas_transport.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    // Get multicomponent properties at several temperatures
+    for (int i = 0; i < ntemps; i++) {
+        temp = 500.0 + 100.0*i;
+        gas->setState_TP(temp, pres);
+        output(0,i) = temp;
+        output(1,i) = multi->viscosity();
+        output(2,i) = multi->thermalConductivity();
+        multi->getThermalDiffCoeffs(&output(3,i));
+    }
```
Comment (bryanwweber) [inline samples/cxx/gas_transport/gas_transport.cpp:71]: Any reason to have 2 separate loops here?

File: src/base/application.cpp [hunk 2]
```diff
@@ -421,36 +422,54 @@ std::string Application::findInputFile(const std::string& name)
...
-            string inname = dirs[i] + "/" + name;
-            std::ifstream fin(inname);
+            string full_name = home + name.substr(1, npos);
+            std::ifstream fin(full_name);
             if (fin) {
-                return inname;
+                return full_name;
+            } else {
+                throw CanteraError("Application::findInputFile",
+                                "Input file '{}' not found", name);
```
Comment (bryanwweber) [inline src/base/application.cpp]: I think the indentation here should line up
```suggestion
                                   "Input file '{}' not found", name);
```

**Review by bryanwweber — CHANGES_REQUESTED** [review]
Looks good, just a few small things. This also means that we no longer need the "Build Samples" step in the CI, right? Since the samples will actually be run during `scons test` now?

File: samples/cxx/gas_transport/gas_transport.cpp (hunk 1 above)
Comment (speth) [reply samples/cxx/gas_transport/gas_transport.cpp:71]: The loops don't have to be separate, but it seems like a simpler example that way, e.g. if you only care about mixture-averaged properties, you can just delete everything after `writePlotFile("transport_mix.csv", ...)`, and then it's easier to see how the second part is just a variation on that theme.

Comment (bryanwweber) [reply samples/cxx/gas_transport/gas_transport.cpp:71]: Yeah, that makes sense. I was kind of torn myself for that reason. Let's leave it as 2 loops, I think that's better as you say.

**Comment by speth** [comment]
> This also means that we no longer need the "Build Samples" step in the CI, right? Since the samples will actually be run during `scons test` now?

Correct. Just pushed a commit that removes these build steps.

**Review by bryanwweber — APPROVED** [review]

**Pull request merged**